src/ray/gcs/gcs_server/gcs_object_manager.cc

```diff
@@ -288,6 +315,12 @@ void GcsObjectManager::LoadInitialData(const EmptyCallback &done) {
   RAY_CHECK_OK(gcs_table_storage_->ObjectTable().GetAll(callback));
 }
 
+void GcsObjectManager::OnJobFinished(const JobID &job_id) {
+  /// Remove caches in object manager first, then items in object table.
+  RemoveAllObjectLocationsInCacheByJobId(job_id);
```

Comment:
This shouldn't be necessary. The ref counting will ensure that all raylets will remove old objects. This is also pretty sketchy since the GCS will become out of sync with the plasma stores.

Reply:
I found the caches will not be remove immediately as test here https://github.com/ray-project/ray/pull/8761/files#diff-b13b9c2b24b42796bde800d10df4f030R603 will fail if we leave the old object alone.

Will dig more to confirm.

Reply:
Hmm but that test is not a full integration test. In the e2e system, the object table will eventually get cleaned up by the raylets.

Actually, one piece of state that does not currently get removed is if the raylet dies. It makes sense for the GCS to clean up the objects that were on the dead raylet.